kubernetes-discovery-client/src/main/java/io/micronaut/kubernetes/configuration/KubernetesConfigMapWatcher.java

```diff
@@ -172,10 +177,16 @@ private void processConfigMapDeleted(ConfigMap configMap) {
         if (passesIncludesExcludesLabelsFilters(configMap)) {
             this.environment.removePropertySource(propertySource);
             KubernetesConfigurationClient.removePropertySourceFromCache(propertySource.getName());
-            this.environment = environment.refresh();
+            refreshEnvironment();
         }
     }
 
+    private void refreshEnvironment() {
+        this.environment.refresh();
+        LOG.trace("Environment changed due to a ConfigMap change, sending RefreshEvent");
+        this.eventPublisher.publishEvent(new RefreshEvent());
```

Comment:
This should call `refreshAndDiff()` and pass the changed keys to the constructor of `RefreshEvent`

This will optimise the beans that get refreshed to only those interested in the changes

Reply:
That's actually what I had implemented initially.

Unfortunately, I think there's a problem with the change detection, because it doesn't return any changes when I try this by editing a ConfigMap.

I actually hat the same problem when using the refresh endpoint to refresh configuration from local files.

Reply:
> That's actually what I had implemented initially.

What I meant to say is that I first tried the following, inspired by what we do in `RefreshController`
```
        final Map<String, Object> changes = this.environment.refreshAndDiff();
        if (!changes.isEmpty()) {
            LOG.trace("Environment changed due to a ConfigMap change, sending RefreshEvent");
            this.eventPublisher.publishEvent(new RefreshEvent(changes));
        }
```
But if you're suggesting we just feed the changes to the `RefreshEvent`, but still send it even if there are "no" changes... then I'm happy to do that.

Reply:
I would like to find out why that doesn't work, it should. 

@pgressa please review the PR

Reply:
@graemerocher now, it does. The problem was the environment was updated _before_ computing the changes.